pkg/tls_extensions/atlsLIstener.go

```diff
@@ -0,0 +1,397 @@
+	handle := cgo.Handle(callbackHandle)
+	defer handle.Delete()
+
+	callback := handle.Value().(FetchAttestation)
+	reportData := C.GoBytes(unsafe.Pointer(reportDataByte), agent.ReportDataSize)
+
+	res := callback(reportData)
+	if res == nil {
+		fmt.Fprintf(os.Stderr, "attestation callback returned nill")
+		return nil
+	}
+
+	*outlen = C.int(len(res))
+	resultC := C.malloc(C.size_t(len(res)))
+	if resultC == nil {
+		fmt.Fprintf(os.Stderr, "could not allocate memory for fetch attestation callback")
+		return nil
+	}
+
+	C.memcpy(resultC, unsafe.Pointer(&res[0]), C.size_t(len(res)))
+
+	return (*C.uchar)(resultC)
+}
+
+type ATLSServerListener struct {
+	tlsListener *C.tls_server_connection
+}
+
+func Listen(addr string, cert []byte, key []byte) (net.Listener, error) {
+	ip, port, err := net.SplitHostPort(addr)
+	if err != nil {
+		return nil, fmt.Errorf("error while creating listener: %v", err)
+	}
+
+	p, err := strconv.Atoi(port)
+	if err != nil {
+		return nil, fmt.Errorf("bad format of IP address: %v", err)
+	}
+
+	cCertPEM := (*C.char)(unsafe.Pointer(&cert[0]))
+	cKeyPEM := (*C.char)(unsafe.Pointer(&key[0]))
+	cIP := C.CString(ip)
+	defer C.free(unsafe.Pointer(cIP))
+
+	atlsListener := C.start_tls_server(
+		cCertPEM, C.int(len(cert)),
+		cKeyPEM, C.int(len(key)),
+		cIP, C.int(p))
+	if atlsListener == nil {
+		return nil, fmt.Errorf("could not create listener")
+	}
+
+	return &ATLSServerListener{tlsListener: atlsListener}, nil
+}
+
+// accept implements the Accept method in the [Listener] interface; it
+// waits for the next call and returns a generic [Conn].
+func (l *ATLSServerListener) Accept() (net.Conn, error) {
+	conn := C.tls_server_accept(l.tlsListener)
+	if conn == nil {
+		return &ATLSConn{tlsConn: nil}, nil
+	}
+
+	return &ATLSConn{tlsConn: conn}, nil
+}
+
+// close stops listening on the TCP address.
+// already Accepted connections are not closed.
+func (l *ATLSServerListener) Close() error {
+	ret := C.tls_server_close(l.tlsListener)
+	if ret != 0 {
+		return fmt.Errorf("could not close the TLS connection")
+	}
+	return nil
+}
+
+// addr returns the listener's network address, a [*TCPAddr].
+// the Addr returned is shared by all invocations of Addr, so
+// do not modify it.
+func (l *ATLSServerListener) Addr() net.Addr {
+	cIP := C.tls_return_addr(&l.tlsListener.addr)
+	defer C.free(unsafe.Pointer(cIP))
+
+	ip := C.GoString(cIP)
+
+	parsedIP := net.ParseIP(ip)
+	if parsedIP == nil {
+		return nil
+	}
+
+	port := C.tls_return_port(&l.tlsListener.addr)
+
+	return &net.TCPAddr{IP: parsedIP, Port: int(port)}
+}
+
+type ATLSConn struct {
+	tlsConn      *C.tls_connection
+	fdReadMutex  sync.Mutex
+	fdWriteMutex sync.Mutex
+	fdDelayMutex sync.Mutex
+}
+
+func (c *ATLSConn) Read(b []byte) (int, error) {
+	c.fdReadMutex.Lock()
+	defer c.fdReadMutex.Unlock()
+
+	if c.tlsConn == nil {
+		return 0, fmt.Errorf("tls connection is NULL")
```

Comment:
```suggestion
		return 0, fmt.Errorf("tls connection is nil")
```
also declare errors using vars

Reply:
Fixed